File: src/main/java/com/flexpoker/game/command/aggregate/TableBalancer.java
package com.flexpoker.game.command.aggregate;

import java.util.Comparator;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import com.flexpoker.exception.FlexPokerException;
import com.flexpoker.game.command.events.PlayerMovedToNewTableEvent;
import com.flexpoker.game.command.events.TablePausedForBalancingEvent;
import com.flexpoker.game.command.events.TableRemovedEvent;
import com.flexpoker.game.command.events.TableResumedAfterBalancingEvent;
import com.flexpoker.game.command.events.GameEvent;

public class TableBalancer {

    private final int maxPlayersPerTable;

    private final UUID gameId;

    public TableBalancer(UUID gameId, int maxPlayersPerTable) {
        this.maxPlayersPerTable = maxPlayersPerTable;
        this.gameId = gameId;
    }

    public Optional<GameEvent> createSingleBalancingEvent(UUID subjectTableId, Set<UUID> pausedTablesForBalancing,
            Map<UUID, Set<UUID>> tableToPlayersMap, Map<UUID, Integer> playerToChipsAtTableMap) {

        // make sure at least two players exists. we're in a bad state if not
        var totalNumberOfPlayers = getTotalNumberOfPlayers(tableToPlayersMap);
        if (totalNumberOfPlayers < 2) {
            throw new FlexPokerException(
                    "the game should be over since less than two players are left.  no balancing should be done");
        }

        // remove any empty tables first
        var emptyTable = tableToPlayersMap.entrySet().stream()
                .filter(x -> x.getValue().isEmpty())
                .findFirst();
        if (emptyTable.isPresent()) {
            return Optional.of(new TableRemovedEvent(gameId, emptyTable.get().getKey()));
        }

        // check to see if the number of tables is greater than the number it
        // should have. if so, move all the players from this table to other
        // tables starting with the one with the lowest number
        if (tableToPlayersMap.size() > getRequiredNumberOfTables(totalNumberOfPlayers)) {
            return createEventToMoveUserFromSubjectTableToAnyMinTable(subjectTableId, tableToPlayersMap,
                    playerToChipsAtTableMap);
        }

        if (isTableOutOfBalance(subjectTableId, tableToPlayersMap)) {
            var tableSize = tableToPlayersMap.get(subjectTableId).size();

            // in the min case, pause the table since it needs to get another
            // player
            if (tableSize == tableToPlayersMap.values().stream()
                    .map(x -> x.size()).min(Integer::compare).get()) {
                return Optional.of(new TablePausedForBalancingEvent(gameId, subjectTableId));
            }

            // only move a player if the out of balance table has the max number
            // of players. an out-of-balance medium-sized table should not send
            // any players
            if (tableSize == tableToPlayersMap.values().stream()
                    .map(x -> x.size()).max(Integer::compare).get()) {
                return createEventToMoveUserFromSubjectTableToAnyMinTable(subjectTableId, tableToPlayersMap,
                        playerToChipsAtTableMap);
            }
        }

        // re-examine the paused tables to see if they are still not balanced
        var pausedTablesThatShouldStayPaused = pausedTablesForBalancing.stream()
                .filter(x -> isTableOutOfBalance(x, tableToPlayersMap))
                .collect(Collectors.toSet());

        // resume the first table that is currently paused, but are now balanced
        var tableToResume = pausedTablesForBalancing.stream()
                .filter(x -> !pausedTablesThatShouldStayPaused.contains(x))
                .findFirst();
        if (tableToResume.isPresent()) {
            return Optional.of(new TableResumedAfterBalancingEvent(gameId, tableToResume.get()));
        }

        return Optional.empty();
    }

    private Optional<GameEvent> createEventToMoveUserFromSubjectTableToAnyMinTable(UUID subjectTableId,
            Map<UUID, Set<UUID>> tableToPlayersMap, Map<UUID, Integer> playerToChipsAtTableMap) {
        var playerToMove = tableToPlayersMap.get(subjectTableId).stream()
                .findFirst().get();
        var toTableId = findNonSubjectMinTableId(subjectTableId, tableToPlayersMap);
        int chips = playerToChipsAtTableMap.get(playerToMove);
        return Optional.of(new PlayerMovedToNewTableEvent(gameId, subjectTableId, toTableId, playerToMove, chips));
    }

    private UUID findNonSubjectMinTableId(UUID subjectTableId,
            Map<UUID, Set<UUID>> tableToPlayersMap) {
        return tableToPlayersMap.entrySet().stream()
                .filter(x -> !x.getKey().equals(subjectTableId))
                .min(setSizeComparator()).get().getKey();
    }

    private boolean isTableOutOfBalance(UUID tableId,
            Map<UUID, Set<UUID>> tableToPlayersMap) {
        var tableSize = tableToPlayersMap.get(tableId).size();

        if (tableSize == 1) {
            return true;
        }

        return tableToPlayersMap.values().stream()
                .anyMatch(x -> x.size() >= tableSize + 2 || x.size() <= tableSize - 2);
    }

    private Comparator<? super Entry<UUID, Set<UUID>>> setSizeComparator() {
        return (x, y) -> Integer.compare(x.getValue().size(), y.getValue().size());
    }

    private int getTotalNumberOfPlayers(Map<UUID, Set<UUID>> tableToPlayersMap) {
        return tableToPlayersMap.values().stream()
                .collect(Collectors.summingInt(Set::size));
    }

    private int getRequiredNumberOfTables(int totalNumberOfPlayers) {
        var numberOfRequiredTables = totalNumberOfPlayers / maxPlayersPerTable;

        if (totalNumberOfPlayers % maxPlayersPerTable != 0) {
            numberOfRequiredTables++;
        }

        return numberOfRequiredTables;
    }

}

File: src/main/kotlin/com/flexpoker/game/command/aggregate/TableBalancer.kt
package com.flexpoker.game.command.aggregate

import com.flexpoker.exception.FlexPokerException
import com.flexpoker.game.command.events.GameEvent
import com.flexpoker.game.command.events.PlayerMovedToNewTableEvent
import com.flexpoker.game.command.events.TablePausedForBalancingEvent
import com.flexpoker.game.command.events.TableRemovedEvent
import com.flexpoker.game.command.events.TableResumedAfterBalancingEvent
import org.pcollections.PMap
import org.pcollections.PSet
import java.util.Optional
import java.util.UUID

fun createSingleBalancingEvent(gameId: UUID, maxPlayersPerTable: Int, subjectTableId: UUID,
                               pausedTablesForBalancing: Set<UUID>, tableToPlayersMap: PMap<UUID, PSet<UUID>>,
                               playerToChipsAtTableMap: Map<UUID, Int>): Optional<GameEvent> {

    // make sure at least two players exists. we're in a bad state if not
    val totalNumberOfPlayers = getTotalNumberOfPlayers(tableToPlayersMap)
    if (totalNumberOfPlayers < 2) {
        throw FlexPokerException(
            "the game should be over since less than two players are left.  no balancing should be done"
        )
    }

    // remove any empty tables first
    val emptyTable = tableToPlayersMap.entries.firstOrNull { it.value.isEmpty() }
    if (emptyTable != null) {
        return Optional.of(TableRemovedEvent(gameId, emptyTable.key))
    }

    // check to see if the number of tables is greater than the number it
    // should have. if so, move all the players from this table to other
    // tables starting with the one with the lowest number
    if (tableToPlayersMap.size > getRequiredNumberOfTables(maxPlayersPerTable, totalNumberOfPlayers)) {
        return createEventToMoveUserFromSubjectTableToAnyMinTable(gameId,
            subjectTableId, tableToPlayersMap, playerToChipsAtTableMap)
    }
    if (isTableOutOfBalance(subjectTableId, tableToPlayersMap)) {
        val tableSize = tableToPlayersMap[subjectTableId]!!.size

        // in the min case, pause the table since it needs to get another player
        if (tableSize == tableToPlayersMap.values.map { it.size }.minOrNull()) {
            return Optional.of(TablePausedForBalancingEvent(gameId, subjectTableId))
        }

        // only move a player if the out of balance table has the max number
        // of players. an out-of-balance medium-sized table should not send
        // any players
        if (tableSize == tableToPlayersMap.values.map { it.size }.maxOrNull()) {
            return createEventToMoveUserFromSubjectTableToAnyMinTable(gameId,
                subjectTableId, tableToPlayersMap, playerToChipsAtTableMap)
        }
    }

    // re-examine the paused tables to see if they are still not balanced
    val pausedTablesThatShouldStayPaused = pausedTablesForBalancing
        .filter { isTableOutOfBalance(it, tableToPlayersMap) }
        .toSet()

    // resume the first table that is currently paused, but are now balanced
    val tableToResume = pausedTablesForBalancing.firstOrNull { !pausedTablesThatShouldStayPaused.contains(it) }

    return if (tableToResume == null) Optional.empty()
    else Optional.of(TableResumedAfterBalancingEvent(gameId, tableToResume))
}

private fun createEventToMoveUserFromSubjectTableToAnyMinTable(gameId: UUID, subjectTableId: UUID,
    tableToPlayersMap: Map<UUID, MutableSet<UUID>>, playerToChipsAtTableMap: Map<UUID, Int>
): Optional<GameEvent> {
    val playerToMove = tableToPlayersMap[subjectTableId]!!.first()
    val toTableId = findNonSubjectMinTableId(subjectTableId, tableToPlayersMap)
    val chips = playerToChipsAtTableMap[playerToMove]!!
    return Optional.of(PlayerMovedToNewTableEvent(gameId, subjectTableId, toTableId, playerToMove, chips))
}

private fun findNonSubjectMinTableId(subjectTableId: UUID, tableToPlayersMap: Map<UUID, MutableSet<UUID>>): UUID {
    return tableToPlayersMap.filter { it.key != subjectTableId }.minByOrNull { it.value.size }!!.key
}

private fun isTableOutOfBalance(tableId: UUID, tableToPlayersMap: Map<UUID, MutableSet<UUID>>): Boolean {
    val tableSize = tableToPlayersMap[tableId]!!.size
    return if (tableSize == 1) {
        true
    } else tableToPlayersMap.values.any { it.size >= tableSize + 2 || it.size <= tableSize - 2 }
}

private fun getTotalNumberOfPlayers(tableToPlayersMap: Map<UUID, MutableSet<UUID>>): Int {
    return tableToPlayersMap.values.sumBy { it.size }
}

private fun getRequiredNumberOfTables(maxPlayersPerTable: Int, totalNumberOfPlayers: Int): Int {
    var numberOfRequiredTables = totalNumberOfPlayers / maxPlayersPerTable
    if (totalNumberOfPlayers % maxPlayersPerTable != 0) {
        numberOfRequiredTables++
    }
    return numberOfRequiredTables
}
